ext/crates/fp/src/limb.rs

```diff
@@ -11,6 +10,146 @@ pub(crate) struct LimbBitIndexPair {
+    /// `end`. Most methods assume that `self.start < entries_per_limb_const::<P>()`, so it advised
+    /// to call `apply_shift` before using the return value.
+    pub(crate) const fn from_start_end(start: usize, end: usize) -> Self {
+        let limb_range = range::<P>(start, end);
+        Self {
+            start,
+            end,
+            limbs: limb_range.end - limb_range.start,
+        }
+    }
+
+    #[inline]
+    pub(crate) const fn limbs(&self) -> usize {
+        self.limbs
+    }
+
+    #[inline]
+    pub(crate) const fn logical(&self) -> usize {
+        self.end - self.start
+    }
+
+    pub(crate) const fn contains(&self, other: &Self) -> bool {
+        self.start + other.end <= self.end
+    }
+
+    /// Shift the entire `LimbLength` backwards so that the start of the range belongs to the first
+    /// limb, and return the number of limbs shifted.
+    pub(crate) const fn apply_shift(&self) -> (Self, usize) {
+        let entries_per = entries_per_limb_const::<P>();
+        let offset = self.start / entries_per;
+        let start = self.start - offset * entries_per;
+        let end = self.end - offset * entries_per;
+        (LimbLength::from_start_end(start, end), offset)
+    }
+
+    pub(crate) const fn restrict_to(&self, other: Self) -> Self {
+        debug_assert!(self.contains(&other));
+        Self::from_start_end(other.start + self.start, other.end + self.start)
+    }
+
+    /// This function panics if `self.start != 0`.
+    pub(crate) fn trim_start(&self, offset: usize) -> Self {
+        debug_assert_eq!(self.start, 0);
+        assert_eq!(offset % entries_per_limb_const::<P>(), 0);
+        let limb_shift = offset / entries_per_limb_const::<P>();
+        Self {
+            start: self.start,
+            end: self.end - offset,
+            limbs: self.limbs - limb_shift,
+        }
+    }
+
+    /// This function assumes that `self.start < entries_per_limb_const::<P>()`. A `LimbLength`
+    /// equivalent to `self` that does satisfy this condition can be obtained by calling
+    /// [`apply_shift`].
+    #[inline]
+    pub(crate) const fn bit_offset(&self) -> usize {
+        self.start * bit_length_const::<P>()
+    }
+
+    #[inline]
+    pub(crate) const fn limb_range(&self) -> Range<usize> {
+        range::<P>(self.start, self.end)
+    }
+
+    /// # Panics
+    ///
+    /// This function underflows if `self.start + self.logical() == 0`, which happens if and only if
+    /// we are taking a slice of width 0 at the start of a limb. This should be a very rare edge
+    /// case. Dealing with the underflow properly would probably require using `saturating_sub` or
+    /// something of that nature, and that has a nontrivial (10%) performance hit.
+    #[inline]
+    pub(crate) fn limb_range_inner(&self) -> Range<usize> {
+        let range = self.limb_range();
+        debug_assert!(range.end > 0, "Underflow");
```

Comment:
I think this can have a more descriptive error message (or none at all; the current one isn't very helpful)

Reply:
Actually I just remembered that overflows and underflows are always checked in debug mode anyway. So we can remove that